Adding a wrapper module with conveniency type classes (#411)
tweag/cooked-validators: README.md

Copyright Tweag I/O 2024

`cooked-validators` is a Haskell library to conveniently and
efficiently write off-chain code for Cardano smart contracts. This
offchain code will be specifically geared to testing and auditing the
smart contract in question with further builtin capabilities of the
library.

> In particular, `cooked-validators` allows to:

Reviewer: ```suggestion
In particular, `cooked-validators` allows the user to:
```

- interact with smart contracts written in Plutus or any other language
  that compiles to [UPLC](https://plutonomicon.github.io/plutonomicon/uplc),
  like for example [Plutarch](https://github.com/Plutonomicon/plutarch-plutus)
  or [Aiken](https://aiken-lang.org/), by loading contracts from byte strings.
- declaratively define transactions in a high level, type-retaining
  data structure, and submitting them for validation, while
  automatically taking care of missing inputs and outputs, balancing,

>   minimum-Ada constraints, collaterals and fees

Reviewer: I find this sentence a bit convoluted and less readable than the previous one. I would rather keep it and add another item about retaining type information in contexts where they are usually lost.

- construct sequences of transactions in an easy-to-understand
  abstraction of "the blockchain", which can be instantiated to
  different actual implementations
- run sequences of transactions in a simulated blockchain
- apply "tweaks" to transactions right before submitting them, where
  "tweaks" are modifications that are aware of the current state of
  the simulated blockchain
- compose and deploy tweaks with flexible idioms inspired by linear
  temporal logic, in order to turn one sequence of transactions into
  many sequences that might be useful test cases, generalized in

>   [Graft](https://github.com/tweag/graft)

Reviewer: I think these were accidentally reformatted and staged in the PR (along with the rest of the text, changed or not, not limited to these few lines).

Author: I will reformat the whole thing with 80 characters, and then we can use a md formatter for later work.

  or [Aiken](https://aiken-lang.org/), by loading contracts from byte strings.
- declaratively define transactions in a high level, type-retaining
  data structure, and submitting them for validation, while
  automatically taking care of missing inputs and outputs, balancing,
  minimum-Ada constraints, collaterals and fees
- construct sequences of transactions in an easy-to-understand
  abstraction of "the blockchain", which can be instantiated to
  different actual implementations
- run sequences of transactions in a simulated blockchain
- apply "tweaks" to transactions right before submitting them, where
  "tweaks" are modifications that are aware of the current state of
  the simulated blockchain
- compose and deploy tweaks with flexible idioms inspired by linear
  temporal logic, in order to turn one sequence of transactions into
  many sequences that might be useful test cases, generalized in
  [Graft](https://github.com/tweag/graft)
- deploy automated attacks over existing sequences of transactions,
  such as datum hijacking or double satisfaction attacks, in an attempt
  to uncover vulnerabilities

> The library is geared specifically towards testing and auditing on-chain code.

Reviewer: Why is this removed?

Author: It is now in the introduction. The sentence just came out of nowhere here.

the `packages` stanza.
   
2. Add the following stanza to the file `cabal.project`
   ```cabal.project
   source-repository-package
     type: git
     location: https://github.com/tweag/cooked-validators
     tag: myTag
     subdir:
       .
   ```

>    where `myTag` is either a commit hash in the repo, or a tag, such as v4.0.0

Reviewer: ```suggestion
   where `myTag` is either a commit hash in the repo, or a tag, such as v4.0.0 (see [available releases]{https://github.com/tweag/cooked-validators/releases}).
```

> ## An example of usage of `cooked-validators`

Reviewer: ```suggestion
### Example
```

   
1. Make your project
   [depend](https://cabal.readthedocs.io/en/stable/getting-started.html#adding-dependencies)
   on `cooked-validators` and `plutus-script-utils`
   

> 3. Enter a Cabal read-eval-print-loop (with `cabal repl`)

Reviewer: ```suggestion
3. Enter a REPL (with `cabal repl`)
```

Author: For some reason, this sentence was there from the beginning so I'm for keeping it this way. That's how I learned what "repl" stands for, maybe others can enjoy this too.

## Documentation

> - The rendered Haddock for the current `main` branch can be found at
> [https://tweag.github.io/cooked-validators/](https://tweag.github.io/cooked-validators/).

Reviewer: ```suggestion
- The rendered Haddock for the current `main` branch can be found
  [here](https://tweag.github.io/cooked-validators/).
```

> - The [CHEATSHEET](doc/CHEATSHEET.md) is a nice entry point and helper
> to keep on sight. It contains many code snippets to quickly get an
> intuition of how to do things. Use it to discover or search for how to
> use features of `cooked-validators`. Note that this is not a tutorial
> nor a ready-to-use recipes book.

Reviewer: The "nice entry point" seems superfluous in retrospect.

```suggestion
- The [CHEATSHEET](doc/CHEATSHEET.md) contains many code snippets to quickly get an
intuition of how to do things. Use it to discover or search for how to
use features of `cooked-validators`. Note that this is not a tutorial
nor a ready-to-use recipes book.
```

> - The [IMPORTS](doc.IMPORTS.md) file is a nice source of information
>   to understand the dependencies we use and how we name them.

Reviewer: ```suggestion
- The [IMPORTS](doc/IMPORTS.md) file describes and helps
  to understand our dependencies and naming conventions for imports.
```

- We also have a
[repository](https://github.com/tweag/cooked-smart-contracts) of
example contracts with offchain code and tests written using

> `cooked-validators`. Note that these examples are not maintained and
> thus written with various versions of the tool.

Reviewer: ```suggestion
`cooked-validators`. Note that some examples are not maintained and
thus written using older versions of cooked-validators.
```

> - Please also look at our
> [issues](https://github.com/tweag/cooked-validators/issues) for
> problems that we're already aware of, and feel free to open new
> issues!

Reviewer: ```suggestion
- Feel free to visit our
[issue tracker](https://github.com/tweag/cooked-validators/issues) to seek help about
known problems, or report new issues!
```
